nexus/operation-ids.md

```diff
@@ -0,0 +1,172 @@
+the caller and the handler. There are two approaches for generating identifiers, caller side and handler side.
+
+### Caller generated identifiers
+
+With this approach, the caller provides an Id to each of the pre-defined RPCs.
+
+The handler _should_ qualify an identifier when dealing with untrusted sources and multi-tenant applications.
+
+The following set of RPCs would be needed to model asynchronous operations using client generated Ids:
+
+- `Start(operation, id, input, callbackURL)` - handler responds inline or indicates that an async operation was started
+- `Cancel(operation, id)` - cancel operation by Id
+- `GetResult(operation, id)` - get operation result by Id
+- `Describe(operation, id)` - get operation status by Id
+- `DeliverResult(token)` - deliver a result of an async operation using a token in the Start request's callback URL
+
+Example invocations:
+
+```js
+Start(
+  'charge',
+  transactionId,
+  { customerId: 'someCustomerId', amount: 5000 },
+  'nexus://caller.example.com/callback?token=' + token
+);
+Cancel('charge', transactionId);
+```
+
+#### Handler addressability
+
+In some cases, the handler might require more than just a single identifier for a given operation, e.g. when an
+operation is backed by a workflow update, which requires both a workflow Id and an update Id. In order to avoid
+maintaining a handler side persistent mapping between caller and handler identifiers, the handler may choose to "leak"
+this detail to the caller, which would need to concatinate multiple identifiers for a given call. The example above
+becomes:
+
+```js
+Start(
+  'charge',
+  customerId + '/' + transactionId,
+  { amount: 5000 },
+  'nexus://caller.example.com/callback?token=' + token
+);
+Cancel('charge', customerId + '/' + transactionId);
+```
+
+> NOTE: This looks a lot like a nested REST style URL, e.g: `/customers/{customerId}/charges/{transactionId}`.
+
+> To take this approach to the extreme, `operation` and `id` could be combined into a single `path` argument, as shown
+> below:
+>
+> - `Start(path, input, callbackURL)`
+> - `Cancel(path)`
+> - `GetResult(path)`
+> - `Describe(path)`
+
+##### Mapping from client to handler identifiers
+
+To avoid leaking the identifier construction logic to the client, the handler would need a mapping of client identifiers
+to handler identifiers. Temporal can provide a strongly consistent store out of the box for this mapping.
+
+A single execution would need to map to multiple client identifiers to account for workflow starts and updates.
+
+Looking up an identifier in persistent storage added latency, and should be opted in to.
+
+### Handler generated identifiers
+
+With this approach, the handler returns a reference Id in response to a start call, which the caller should persist for
+use with the pre-defined RPCs as shown below:
+
+- `Start(operation, input, callbackURL, cancelBeforeStarted)` - handler responds inline or returns a reference Id
+  ([`*`](#cancel-before-started))
```

Comment:
We need to add a client generated transaction id to `Start(transaction_id,...)` to allow "exactly-once" semantics for the operations. This transaction id may, or may not, ended up as part of the handler id, this is implementation dependent and not visible to the client (but may simplify the handler implementation...). The other calls do not need the transactional ID because they are naturally idempotent, or side effect free...

Reply:
Transaction ID would be part of the input here. The handler would need to interpret it as an idempotency token.